app/controllers/v3/application_controller.rb

```diff
@@ -90,7 +90,19 @@ def parsed_yaml
     return @parsed_yaml if @parsed_yaml
 
     allow_yaml_aliases = false
-    yaml = Psych.safe_load(request.body.string, permitted_classes: [], permitted_symbols: [], aliases: allow_yaml_aliases, strict_integer: true)
+
+    # Thin and also Puma outsource large request bodies/headers to a tempfile.
+    # So we need to check here whether it is a StringIO or a Tempfile and load it accordingly.
+    if request.body.instance_of? Tempfile
+      if request.body.size > 1.megabyte
+        bad_request!('Manifest size is too large. The maximum supported size is 1MB.')
+      else
+        yaml = Psych.load_file(request.body, permitted_classes: [], permitted_symbols: [], aliases: allow_yaml_aliases, strict_integer: true)
+      end
+    else
+      yaml = Psych.safe_load(request.body.string, permitted_classes: [], permitted_symbols: [], aliases: allow_yaml_aliases, strict_integer: true)
```

Comment:
You can use `request.body.read` in both cases, so the `if-else` is not necessary. The check for the size still makes sense and should be done before.